CDCgov/prime-simplereport
# Johanna/5846 manage users no facility user type

# BACKEND AND FRONTEND PULL REQUEST

## Related Issue

- Resolves #5846 

## Changes Proposed

The error message that needed to be addressed in the ticket is displayed by a global logic in the app that surfaces any graphQL error, so in order to stop displaying the message the fix was performed in the backend.

Root cause seems to be the way that we were obtaining the variable `accessibleFacilities` which contains the facilities that a user has access to. The edge case was having a user with a list of claims that included facilityIds of soft deleted facilities. The `accessibleFacilities` would still contain an array with a number of objects matching the number of facilities in the `claims` but the objects were null. These null instances were triggering the exception in the graphql endpoint. I just reworked the way we were obtaining the `accessibleFacilities` set to just do a filtering based on whether or not the facilityId was part of the `claims` set. 

I also added a warning in `user details` when no facilities are returned in the `testingFacilities` array.
<img width="926" alt="Screenshot 2023-06-22 at 9 18 24 AM" src="https://github.com/CDCgov/prime-simplereport/assets/103958711/c3dc5aae-2420-4806-b806-92c5aef98585">

## Additional Information
While working on the ManageUsers component I noticed that we were performing duplicated calls to the getUser query very often. Root cause was that the logic to trigger the call relied on states and useEffects that executed too soon or too late while triggering the query, so apollo client would react to the state updates in an odd way and trigger the query again.

## Testing

Code is available in [dev2](https://dev2.simplereport.gov/)

- Create a new facility
- Assign a user to only that new facility
- Delete the facility you created
- Verify that the user loads normally but displays the warning.
- Please also do regression testing on add a user and delete a user!.
